File: go/backtracking/51_solve_n_queens.go
package backtracking

import (
	"strings"
)

// solveNQueens N皇后
func solveNQueens(n int) [][]string {

	// 搜索中满足条件的结果
	path := make([][]string, n)
	for i := 0; i < n; i++ {
		path[i] = make([]string, n)
		for j := 0; j < n; j++ {
			path[i][j] = "."
		}
	}

	// 所有满足条件的结果集
	res := make([][]string, 0)

	// 回溯算法
	var backtrack func(row int)
	backtrack = func(row int) {
		// 遍历到最后一行了
		if row == n {
			tmp := make([]string, 0)
			for i := 0; i < n; i++ {
				tmp[i] = strings.Join(path[i], "")
			}
			res = append(res, tmp)
			return
		}

		for col := 0; col < n; col++ {
			if isValid(row, col, n, path) {
				path[row][col] = "Q"
				backtrack(row + 1)
				// 回溯
				path[row][col] = "."
			}
		}
	}

	backtrack(0)
	return res
}

// isValid path[r][c] 是否可以设置为Q
// 同行，同列，对角线的都为无效
func isValid(r, c int, n int, path [][]string) bool {

	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			if path[i][j] == "Q" && (j == c || i+j == r+c || i-j == r-c) {
				return false
			}
		}
	}
	return true
}

File: rust/src/backtracking/_51_solve_n_queens.rs
/// N 皇后
pub fn solve_n_queens(n: i32) -> Vec<Vec<String>> {
    let n = n as usize;
    // 初始化棋盘, 同时作为搜索中满足条件的path
    let mut path = vec![vec!["."; n]; n];

    // 搜索结果
    let mut res = vec![];

    fn backtrack(res: &mut Vec<Vec<String>>, path: &mut Vec<Vec<&str>>, row: usize, n: usize) {
        // return condition
        if row == n {
            let mut p = vec![];
            for item in path {
                p.push(item.join(""));
            }
            res.push(p);
            return;
        }

        for col in 0..n {
            if is_valid(path, row, col, n) {
                path[row][col] = "Q";
                backtrack(res, path, row + 1, n);
                path[row][col] = ".";
            }
        }
    }

    backtrack(&mut res, &mut path, 0, n);
    res

}

/// 判断path[row][col] 是否可以放置Q
/// 同行、同列、对接线不可放置
fn is_valid(path: &Vec<Vec<&str>>, row: usize, col: usize, n: usize) -> bool {
    for i in 0..n {
        for j in 0..n {
            if path[i][j] == "Q" && (j == col || i + j == row + col || i - j == row - col) {
                return false;
            }
        }
    }
    true
}
